semantic-analysis/src/Analysis/Effect/Domain.hs

```diff
@@ -39,10 +39,10 @@ dvar = send . DVar
 
 -- Functions
 
-dabs :: Has (Dom val) sig m => Name -> (val -> m val) -> m val
+dabs :: Has (Dom val) sig m => [Name] -> ([val] -> m val) -> m val
 dabs n = send . DAbs n
 
-dapp :: Has (Dom val) sig m => val -> val -> m val
+dapp :: Has (Dom val) sig m => val -> [val] -> m val
```

Comment:
As described above, this is the key place where n-ary functions/applications are required.